### PR TITLE
No-Op runner

### PR DESCRIPTION
**Pull request by jfryman**
This PR introduces a runner that does *absolutely nothing*. Yup!

I need this as a stub and placeholder when all I need to do is send a notification, but it's inappropriate to do in the context of the task I am currently in.

:eyes: and other critiques are most certainly appreciated.

### CODE REVIEW

**Comment by manasdk** [comment]
> I need this as a stub and placeholder when all I need to do is send a notification, but it's inappropriate to do in the context of the task I am currently in.

Can you provide an example?

**Comment by jfryman** [comment]
@manasdk sure.

```
---
chain:
  -
    name: "search_kb"
    ref: "reamaze.article_search"
    params:
      query: "{{ query }}"
    publish:
      kb_results: "{{ search_kb.result }}"
    on-success: "chatops_parse"
    on-failure: "fail_search_kb"
  -
    name: "chatops_parse"
    ref: "st2-community.parse"
    params:
      input: "{{ kb_results }"
    publish:
      seach_chatops_output: "{{ result }}"
  -
    name: "send_chatops_to_user"
    ref: "core.noop"
    params: {}
    notify:
      on-success:
        message: "Found some articles for you! Check them out! {{ kb_results }}"   
  -
    name: "fail_search_kb"
    ref: "core.noop"
    params: {}
    notify:
      on-success:
        message: "Unable to find KB article for {{query}}"
```

In action chain, it's less prevalent to see, but I see a notification coupled to the action that is performing something to be reserved for system notifications, but having a `noop` lets me keep human responses in a separate leaf in my workflow.

**Comment by manasdk** [comment]
There is perhaps some nuance here that I am missing why the notification should be separate. 

In any case a `noop` action seems reasonable - it is unfortunate that it requires a special runner as well but it is what it is I suppose.

File: st2actions/st2actions/runners/nooprunner.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    KEYS_TO_TRANSFORM = ['stdout', 'stderr']
+
+    def __init__(self, runner_id):
+        super(NoopRunner, self).__init__(runner_id=runner_id)
+
+    def run(self, action_parameters):
+
+        LOG.info('Executing action via NoopRunner: %s', self.runner_id)
+        LOG.info('[Action info] name: %s, Id: %s' %
+                 (self.action_name, str(self.liveaction_id)))
```
Comment (manasdk) [inline st2actions/st2actions/runners/nooprunner.py]: should use execution_id.

File: st2actions/st2actions/runners/nooprunner.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    KEYS_TO_TRANSFORM = ['stdout', 'stderr']
+
+    def __init__(self, runner_id):
+        super(NoopRunner, self).__init__(runner_id=runner_id)
+
+    def run(self, action_parameters):
+
+        LOG.info('Executing action via NoopRunner: %s', self.runner_id)
+        LOG.info('[Action info] name: %s, Id: %s' %
```
Comment (manasdk) [inline st2actions/st2actions/runners/nooprunner.py]: `LOG.info('[Action info] name: %s, Id: %s', self.action_name, str(self.liveaction_id))`

is better as the evaluation of the string gets delayed.

File: st2actions/st2actions/runners/nooprunner.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                 (self.action_name, str(self.liveaction_id)))
+
+        result = {
+            'failed': False,
+            'succeeded': True,
+            'return_code': 0,
+            'stdout': "",
```
Comment (manasdk) [inline st2actions/st2actions/runners/nooprunner.py]: Is there much value in including `stdout` and `stderr` properties? If not I would skip them for the no-op runner

File: st2actions/st2actions/runners/nooprunner.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        super(NoopRunner, self).__init__(runner_id=runner_id)
+
+    def run(self, action_parameters):
+
+        LOG.info('Executing action via NoopRunner: %s', self.runner_id)
+        LOG.info('[Action info] name: %s, Id: %s' %
+                 (self.action_name, str(self.liveaction_id)))
+
+        result = {
+            'failed': False,
```
Comment (manasdk) [inline st2actions/st2actions/runners/nooprunner.py]: (unrelated)
It has always struck me as odd that we have 2 separate properties `failed` and `succeeded`.

File: st2common/st2common/bootstrap/runnersregistrar.py [hunk 5]
```diff
@@ -521,6 +521,14 @@
             }
         },
         'runner_module': 'st2actions.runners.cloudslang.cloudslang_runner'
+    },
+    {
+        'name': 'noop',
+        'aliases': [],
+        'description': 'A runner that does absolutely nothing',
```
Comment (manasdk) [inline st2common/st2common/bootstrap/runnersregistrar.py]: Now that not true :)

It actually simply responds with a static response which does not change regardless of the input.

File: st2actions/st2actions/runners/nooprunner.py (hunk 3 above)
Comment (jfryman) [inline st2actions/st2actions/runners/nooprunner.py]: No, none at all. What I was more concerned about is if this is consumed somewhere else upstream... so I kept the payload consistent. I'll drop if that's safe

Comment (manasdk) [inline st2actions/st2actions/runners/nooprunner.py]: Drop it. Runners are allowed to define the structure of the result and that is generally treated as opaque by StackStorm. 

Workflows and rules tend to be runner aware as they need to intrinsically understand these properties to consume the result. There should be no impact in this case so I would keep the response specific to the no-op runner.

File: st2actions/st2actions/runners/nooprunner.py (hunk 4 above)
Comment (jfryman) [inline st2actions/st2actions/runners/nooprunner.py]: Yeah, I'll express the same puzzlement. status: Failed/Success.

If any historians want to throw down some knowledge, always fun to know. :)

**Comment by manasdk** [comment]
Please update the CHANGELOG.

**Comment by jfryman** [comment]
:green_apple:

Gonna merge! Thanks for the :eyes:, @manasdk!

**Comment by dzimine** [comment]
This is awesome! 
I was thinking about something like this to do a "dry-run". Just to beef this up with runner parameters "dry-result" and "dry-status", and we can create any test workflows. Further, we can add these parameters to a base runner, make other runners ignore it and have a dryrun key, so that when it's used, the dry-runner is substituted for all other runners, and actions return dry resutl and dry status. Again this is a rough idea, you guys can make something good from it, hopefully. 

But this is awesome as it is, for testing workflows.

**Comment by Kami** [comment]
Yeah, `noop` action is something we have discussed before - there are definitely some valid uses in the workflows.

As far as the runner goes - couldn't we simple make it a `core.local` action which does `cmd=exit 0`"? I know it's not exactly the same as doing nothing, but it's close enough :P

File: st2actions/st2actions/runners/nooprunner.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    def run(self, action_parameters):
+
+        LOG.info('Executing action via NoopRunner: %s', self.runner_id)
+        LOG.info('[Action info] name: %s, Id: %s',
+                 self.action_name, str(self.execution_id))
+
+        result = {
+            'failed': False,
+            'succeeded': True,
+            'return_code': 0,
```
Comment (Kami) [inline st2actions/st2actions/runners/nooprunner.py]: `return_code` is specific to local and remote runners so we can also omit it here. Same goes for `failed` attribute.

**Pull request merged**